### PR TITLE
Refactor resources to produce less `ResourcesChanged` events.

### LINKED ISSUE #3947: Too many ResourcesChanged events are raised.
Opening an empty `Window` causes 636 instances of `ResourcesChangedEventArgs` to be created, meaning that `ResourcesChanged` is being raised that many times.

Opening `ControlCatalog` on the first page causes the event to be raised 3032 times! This feels rather excessive and is definitely not good for perf.

### PR DESCRIPTION
**Pull request by grokys**
## What does the pull request do?

Reduces the number of `ResourcesChanged` events raised and refactors the resource code to make a bit more sense.

Split resource nodes into two types:

- `IResourceHost` represents controls and `Application`
- `IResourceProvider` represents resource dictionaries and styles, these are owned by `IResourceHost`s

Dynamic resources are now always resolved from an `IResourceHost`: if an `IResourceProvider` doesn't have a host then a dynamic resource in the resource provider will not be resolved. Resource providers no longer have a `ResourcesChanged`  event, instead they notify their `IResourceHost` owner of a resource change by calling the `NotifyHostedResourcesChanged` method.

Instead of each control listening to its parent `ResourcesChanged` event, use the logical tree to propagate `ResourcesChanged` events:

- When attaching to the logical tree, piggyback on the `AttachedToLogicalTree` traversal
- At other times, propagate by calling `NotifyResourcesChanged`

`ResourcesChanged` event reduction:

- Empty window from 636 to 0
- ControlCatalog from 3032 to 594

## Checklist

- [x] Added unit tests (if possible)?

## Fixed issues

Fixes #3947

### CODE REVIEW

File: src/Avalonia.Styling/Styling/Styles.cs [hunk 1]
```diff
@@ -54,43 +63,53 @@ public IResourceDictionary Resources
...
+        {
+            get
+            {
+                if (Count > 0)
+                {
+                    return true;
+                }
+
+                foreach (var i in this)
+                {
```
Comment (MarchingCube) [inline src/Avalonia.Styling/Styling/Styles.cs:92]: We get here only when Count is 0, so wouldn’t this loop be a no-op?

Comment (grokys) [reply src/Avalonia.Styling/Styling/Styles.cs:92]: Yep, brainfart here. The `Count` check can be removed.

Comment (grokys) [reply src/Avalonia.Styling/Styling/Styles.cs:92]: Oh, no, actually it was supposed to be checking `_resources.Count`.

**Review by MarchingCube — APPROVED** [review]
LGTM.

**Pull request merged**